### PR TITLE
elementwise broadcast function enhancement

### PR DESCRIPTION
**Pull request by danleifeng**
elementwise broadcast function enhancement:
- support broadcast operation with ```m*n*k m*1*k``` dimensions. #DLTP-256

add unit tests of new broadcast type for all elementwise operations.

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/ngraph/test_elementwise_max_ngraph_op.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 import unittest, sys
 sys.path.append("../")
-from test_elementwise_max_op import *
-
+# from test_elementwise_max_op import *
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/ngraph/test_elementwise_max_ngraph_op.py]: Clean up this and so on the downside.

File: python/paddle/fluid/tests/unittests/ngraph/test_elementwise_min_ngraph_op.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 
 import unittest, sys
 sys.path.append("../")
-from test_elementwise_min_op import *
-
+# from test_elementwise_min_op import *
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/ngraph/test_elementwise_min_ngraph_op.py]: Clean up this!

File: paddle/fluid/operators/elementwise/elementwise_op_function.h [hunk 3]
```diff
@@ -50,22 +50,56 @@ namespace operators {
  */
 inline void get_mid_dims(const framework::DDim &x_dims,
                          const framework::DDim &y_dims, const int axis,
-                         int *pre, int *n, int *post) {
+                         int *pre, int *n, int *post, int *mid_flag = NULL) {
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_op_function.h:59]: Add a comment of mid_flag.
And add an example of it in the above comment.

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/tests/unittests/ngraph/test_elementwise_max_ngraph_op.py (hunk 1 above)
Comment (danleifeng) [reply python/paddle/fluid/tests/unittests/ngraph/test_elementwise_max_ngraph_op.py]: Done.

File: python/paddle/fluid/tests/unittests/ngraph/test_elementwise_min_ngraph_op.py (hunk 2 above)
Comment (danleifeng) [reply python/paddle/fluid/tests/unittests/ngraph/test_elementwise_min_ngraph_op.py]: Done.

File: paddle/fluid/operators/elementwise/elementwise_op_function.h (hunk 3 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_op_function.h:59]: Done.

File: paddle/fluid/operators/elementwise/elementwise_op_function.h [hunk 4]
```diff
@@ -533,23 +663,38 @@ void ElemwiseGradComputeWithBroadcast(
...
+  int pre, n, post, mid_flag;
+  get_mid_dims(x_dim, y_dim, axis, &pre, &n, &post, &mid_flag);
+  if (mid_flag) {
+    if (platform::is_gpu_place(ctx.GetPlace())) {
+#ifdef __NVCC__
+      ElemwiseGradBroadcastMid2CUDA(
+          ctx.template device_context<DeviceContext>().stream(), x.data<T>(),
+          y.data<T>(), out.data<T>(), dout.data<T>(), pre, n, post, dx_op,
+          dy_op, dx == nullptr ? nullptr : dx->mutable_data<T>(ctx.GetPlace()),
+          dy == nullptr ? nullptr : dy->mutable_data<T>(ctx.GetPlace()));
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_op_function.h:676]: #else

File: paddle/fluid/operators/elementwise/elementwise_op_function.h [hunk 5]
```diff
@@ -533,23 +663,38 @@ void ElemwiseGradComputeWithBroadcast(
...
+          dx == nullptr ? nullptr : dx->mutable_data<T>(ctx.GetPlace()),
+          dy == nullptr ? nullptr : dy->mutable_data<T>(ctx.GetPlace()));
+    }
+  } else if (post == 1) {
     if (platform::is_gpu_place(ctx.GetPlace())) {
 #ifdef __NVCC__
       ElemwiseGradBroadcast1CUDA(
           ctx.template device_context<DeviceContext>().stream(), x.data<T>(),
-          y.data<T>(), out.data<T>(), dout.data<T>(), h, w, dx_op, dy_op,
+          y.data<T>(), out.data<T>(), dout.data<T>(), pre, n, dx_op, dy_op,
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_op_function.h:690]: #else

File: paddle/fluid/operators/elementwise/elementwise_op_function.h [hunk 6]
```diff
@@ -47,25 +47,65 @@ namespace operators {
...
                          const framework::DDim &y_dims, const int axis,
-                         int *pre, int *n, int *post) {
+                         int *pre, int *n, int *post, int *mid_flag = NULL) {
   *pre = 1;
   *n = 1;
   *post = 1;
-  for (int i = 0; i < axis; ++i) {
-    (*pre) *= x_dims[i];
-  }
+  if (mid_flag != NULL) {
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_op_function.h:63]: check only dim=1

File: paddle/fluid/operators/elementwise/elementwise_op_function.h [hunk 7]
```diff
@@ -533,23 +663,38 @@ void ElemwiseGradComputeWithBroadcast(
   auto y_dim = trim_trailing_singular_dims(y_dim_untrimed);
   axis = (y_dim.size() == 0) ? x_dim.size() : axis;
 
-  int pre, n, post;
-  get_mid_dims(x_dim, y_dim, axis, &pre, &n, &post);
-  if (post == 1) {
-    int h = pre;
-    int w = n;
+  int pre, n, post, mid_flag;
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_op_function.h]: mid_flag=0

File: paddle/fluid/operators/elementwise/elementwise_op_function.h [hunk 8]
```diff
@@ -533,23 +663,38 @@ void ElemwiseGradComputeWithBroadcast(
...
+  get_mid_dims(x_dim, y_dim, axis, &pre, &n, &post, &mid_flag);
+  if (mid_flag) {
+    if (platform::is_gpu_place(ctx.GetPlace())) {
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_op_function.h:670]: PADDLE_ENFORCE_EQ(mid_flag,1)

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/elementwise/elementwise_op_function.h (hunk 6 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_op_function.h:63]: Done. (72)

File: paddle/fluid/operators/elementwise/elementwise_op_function.h (hunk 7 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_op_function.h]: Done.

File: paddle/fluid/operators/elementwise/elementwise_op_function.h (hunk 8 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_op_function.h:670]: Done.

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**